src/main/scala/analysis/IrreducibleLoops.scala

```diff
@@ -55,47 +51,138 @@ class Loop(val header: CFGPosition) {
 /* Loop detection and classification with respect to being reducible or irreducible. Implements the algorithm
  *  described by Wei in `A New Algorithm for Identifying Loops in Decompilation` (LNCS 4632 pp 170-183)
  */
-class LoopDetector(cfg: Program) {
-  // Header -> Loop
-  private val loops: mutable.HashMap[CFGPosition, Loop] = mutable.HashMap()
-  private val headers: mutable.Set[CFGPosition] = mutable.Set()
+object LoopDetector {
 
-  // Algorithm helpers
-  private val visitedNodes: mutable.Set[CFGPosition] = mutable.Set()
-  private val nodeDFSPpos: mutable.HashMap[CFGPosition, Int] = mutable.HashMap()
-  private val iloopHeaders: mutable.HashMap[CFGPosition, CFGPosition] = mutable.HashMap()
-  private val edgeStack: mutable.Stack[LoopEdge] = mutable.Stack()
+  case class State(
+    // Header -> Loop
+    loops: Map[Block, Loop] = Map(),
+    headers: Set[Block] = Set(),
 
-  /*
-   * Returns the set of irreducible loops in the program.
-   *
-   */
-  def irreducible_loops(): Set[Loop] = {
-    val irreducibleLoops: Set[Loop] = loops.values.filter(l => !l.reducible).toSet
-    irreducibleLoops
-
-    /*
-    TODO is this supposed to be anything?
-    if (irreducibleLoops.isEmpty) {
-      irreducibleLoops
-    } else {
-      val wantedLoops: mutable.Set[Loop] = mutable.Set[Loop]()
-      val irrHeaders: Set[CFGPosition] = irreducibleLoops.map(l => l.header)
+    // Algorithm helpers
+    visitedNodes: Set[Block] = Set(),
+    nodeDFSPpos: Map[Block, Int] = Map(),
+    iloopHeaders: Map[Block, Block] = Map(),
+    edgeStack: List[LoopEdge] = List(),
+
+  ) {
+    def irreducibleLoops = loops.values.filter(l => !l.reducible).toSet
+
+    def identifiedLoops = loops.values
 
-      irreducibleLoops
+    def reducibleTransformIR() : State = {
+      this.copy(loops = LoopTransform.llvm_transform(loops.values).map(l => l.header -> l).toMap)
     }
-    */
+  }
+
+  def identify_loops(entryBlock: Block): State = {
+    traverse_loops_dfs(State(), entryBlock, 1)._1
   }
 
   /*
-   * Returns the set of loops in the program.
+   * Returns the set of loops for each procedure in the program.
    */
-  def identify_loops(): Set[Loop] = {
-    val funcEntries = cfg.procedures
-    funcEntries.foreach { funcEntry => traverse_loops_dfs(funcEntry, 1) }
-    loops.values.toSet
+  def identify_loops(cfg: Program): State = {
+    cfg.procedures.toSet.map(_.entryBlock).flatten
+      .foldLeft(State())((st, eb) => (traverse_loops_dfs(st, eb, 1)._1))
   }
 
+
+  private def processVisitedNodeOutgoingEdge(istate: State, edge: LoopEdge, DFSPpos: Int): State = {
```

Comment:
DFSPpos is never used, is there a reason it exists?

Reply:
I think I just missed it in the refactor, its been moved into State.